File: boundedblockingqueue.go
package main

import (
    "fmt"
    "sync"
)

// BoundedBlockingQueue is a struct that implements a bounded blocking queue
type BoundedBlockingQueue struct {
    size int
    queue []interface{}
    mutex sync.Mutex
    notEmpty sync.Cond
    notFull sync.Cond
}

// NewBoundedBlockingQueue returns a new instance of BoundedBlockingQueue
func NewBoundedBlockingQueue(size int) *BoundedBlockingQueue {
    queue := &BoundedBlockingQueue{size: size}
    queue.notEmpty = sync.Cond{L: &queue.mutex}
    queue.notFull = sync.Cond{L: &queue.mutex}
    return queue
}

// Put adds an element to the queue
func (queue *BoundedBlockingQueue) Put(element interface{}) {
    queue.mutex.Lock()
    defer queue.mutex.Unlock()
    for queue.size == len(queue.queue) {
        queue.notFull.Wait()
    }
    queue.queue = append(queue.queue, element)
    queue.notEmpty.Signal()
}

// Take removes and returns the first element from the queue
func (queue *BoundedBlockingQueue) Take() interface{} {
    queue.mutex.Lock()
    defer queue.mutex.Unlock()
    for len(queue.queue) == 0 {
        queue.notEmpty.Wait()
    }
    element := queue.queue[0]
    queue.queue = queue.queue[1:]
    queue.notFull.Signal()
    return element
}

func main() {
    queue := NewBoundedBlockingQueue(2)
    queue.Put(1)
    queue.Put(2)
    fmt.Println(queue.Take())
    fmt.Println(queue.Take())
}

File: boundedblockingqueue.rs
use std::sync::{Arc, Mutex, Condvar};
use std::thread;

struct BoundedBlockingQueue {
    size: usize,
    queue: Vec<i32>,
    lock: Mutex<()>,
    not_empty: Condvar,
    not_full: Condvar,
}

impl BoundedBlockingQueue {
    fn new(size: usize) -> Arc<Self> {
        let queue = Arc::new(BoundedBlockingQueue {
            size,
            queue: Vec::new(),
            lock: Mutex::new(()),
            not_empty: Condvar::new(),
            not_full: Condvar::new(),
        });

        queue
    }

    fn put(&self, element: i32) {
        let mut guard = self.lock.lock().unwrap();
        while self.queue.len() == self.size {
            guard = self.not_full.wait(guard).unwrap();
        }
        self.queue.push(element);
        self.not_empty.notify_one();
    }

    fn take(&self) -> i32 {
        let mut guard = self.lock.lock().unwrap();
        while self.queue.is_empty() {
            guard = self.not_empty.wait(guard).unwrap();
        }
        let element = self.queue.remove(0);
        self.not_full.notify_one();
        element
    }
}

fn main() {
    let queue = BoundedBlockingQueue::new(2);
    queue.put(1);
    queue.put(2);
    println!("{}", queue.take());
    println!("{}", queue.take());
}
